File: relay/relay.js
#!/usr/bin/env node
/* BrowserOS Virtual Switch Relay (Node.js)
 * ===========================================================================
 * Gives multiple browser-tab VMs a REAL multi-VM network:
 *   - each connected VM is a port on a virtual Layer-2 switch
 *   - Ethernet frames are forwarded between VMs  -> they can ping/SSH each other
 *   - a built-in DHCP server hands each VM a UNIQUE IP (10.5.0.2, .3, .4, ...)
 *   - built-in ARP + (optional) NAT toward the host's internet
 *
 * v86's "simple" network adapter speaks raw Ethernet frames over a binary
 * WebSocket (binaryType=arraybuffer). This server implements that protocol.
 *
 * Run:   node relay.js            (listens on ws://127.0.0.1:9000)
 * LAN:   BROWSEROS_ALLOW_LAN=1 node relay.js
 * Then in BrowserOS: Network -> NAT Network, Relay URL = ws://<host-ip>:9000
 *
 * NOTE: NAT to the internet uses a raw socket and requires elevated
 * privileges + a host with raw packet support. Inter-VM comms + DHCP work
 * everywhere; full internet NAT is best-effort (see README in this folder).
 */
"use strict";
const http = require("http");
const crypto = require("crypto");

let WebSocketServer;
try { WebSocketServer = require("ws").Server; }
catch (e) {
  console.error("Missing locked dependency 'ws'. Install it with:  npm ci");
  process.exit(1);
}

const PORT = parseInt(process.env.PORT || "9000", 10);
const ALLOW_LAN = /^(1|true|yes|on)$/i.test(process.env.BROWSEROS_ALLOW_LAN || "");
const HOST = ALLOW_LAN ? "0.0.0.0" : "127.0.0.1";
const MAX_FRAME = 65536;
const MAX_CLIENTS = 64;
const MAX_BUFFERED = 1024 * 1024;
const ALLOWED_ORIGINS = new Set(
  (process.env.BROWSEROS_ALLOWED_ORIGINS || "")
    .split(",").map(s => s.trim()).filter(Boolean)
);
const NET = "10.5.0";          // virtual subnet 10.5.0.0/24
const GW  = NET + ".1";        // gateway / relay address
const NETMASK = "255.255.255.0";
const DNS = "10.5.0.1";        // relay answers/forwards DNS (best effort)
const LEASE = 86400;

// ---- helpers ----
const ip2buf = (s) => Buffer.from(s.split(".").map(Number));
const buf2ip = (b, o = 0) => `${b[o]}.${b[o+1]}.${b[o+2]}.${b[o+3]}`;
const mac2str = (b, o = 0) => [...b.slice(o, o+6)].map(x=>x.toString(16).padStart(2,"0")).join(":");
function checksum(buf, start, end) {
  let sum = 0;
  for (let i = start; i < end; i += 2) sum += (buf[i] << 8) + (buf[i+1] || 0);
  while (sum >> 16) sum = (sum & 0xffff) + (sum >> 16);
  return (~sum) & 0xffff;
}
function hostName(host) {
  try { return new URL("http://" + host).hostname.replace(/^\[|\]$/g, "").toLowerCase(); }
  catch (_) { return ""; }
}
function trustedHost(host) {
  if (host === "localhost" || host === "127.0.0.1" || host === "::1") return true;
  const p = host.split(".").map(Number);
  if (p.length !== 4 || p.some(n => !Number.isInteger(n) || n < 0 || n > 255)) return false;
  return ALLOW_LAN && (
    p[0] === 10 ||
    p[0] === 127 ||
    (p[0] === 169 && p[1] === 254) ||
    (p[0] === 172 && p[1] >= 16 && p[1] <= 31) ||
    (p[0] === 192 && p[1] === 168)
  );
}
function trustedOrigin(req) {
  const origin = req.headers.origin || "";
  if (ALLOWED_ORIGINS.has(origin)) return true;
  let u;
  try { u = new URL(origin); } catch (_) { return false; }
  if (u.protocol !== "http:" && u.protocol !== "https:") return false;
  const originHost = u.hostname.replace(/^\[|\]$/g, "").toLowerCase();
  return originHost === hostName(req.headers.host || "") && trustedHost(originHost);
}
function safeSend(ws, payload) {
  if (ws.readyState !== 1) return false;
  if (ws.bufferedAmount > MAX_BUFFERED) {
    ws.terminate();
    return false;
  }
  ws.send(payload);
  return true;
}

// ---- switch state ----
let nextHost = 2;
const clients = new Map();      // ws -> { mac, ip }
const macTable = new Map();     // macStr -> ws  (learned)
const gwMac = Buffer.from([0x02, 0x00, 0x00, 0x00, 0x00, 0x01]);

function allocIP() {
  const ip = `${NET}.${nextHost}`;
  nextHost = nextHost >= 250 ? 2 : nextHost + 1;
  return ip;
}

// ---- DHCP server (replies to DISCOVER/REQUEST) ----
function handleDHCP(ws, frame) {
  // Ethernet(14) + IP(20) + UDP(8) + BOOTP/DHCP
  const ethType = (frame[12] << 8) | frame[13];
  if (ethType !== 0x0800) return false;            // not IPv4
  const ihl = (frame[14] & 0x0f) * 4;
  const proto = frame[23];
  if (proto !== 17) return false;                  // not UDP
  const udp = 14 + ihl;
  const srcPort = (frame[udp] << 8) | frame[udp+1];
  const dstPort = (frame[udp+2] << 8) | frame[udp+3];
  if (dstPort !== 67) return false;                // not DHCP
  const dhcp = udp + 8;
  if (dhcp + 240 > frame.length) return false;
  const xid = frame.slice(dhcp+4, dhcp+8);
  const chaddr = frame.slice(dhcp+28, dhcp+34);    // client MAC
  // find DHCP message type (option 53)
  let opt = dhcp + 240, msgType = 0, reqIp = null;
  while (opt < frame.length && frame[opt] !== 255) {
    if (frame[opt] === 0) { opt++; continue; }
    if (opt + 1 >= frame.length) return false;
    const code = frame[opt], len = frame[opt+1];
    if (opt + 2 + len > frame.length) return false;
    if (code === 53) msgType = frame[opt+2];
    if (code === 50) reqIp = frame.slice(opt+2, opt+6);
    opt += 2 + len;
  }
  const cli = clients.get(ws);
  if (!cli.ip) cli.ip = allocIP();
  const yourIp = ip2buf(cli.ip);
  const replyType = msgType === 1 ? 2 : 5;         // DISCOVER->OFFER, REQUEST->ACK

  // Build DHCP reply
  const dhcpLen = 240 + 3+3 + 6 + 6 + 6 + 6 + 6 + 1; // options below
  const out = Buffer.alloc(14 + 20 + 8 + 300);
  // Ethernet
  chaddr.copy(out, 0); gwMac.copy(out, 6); out[12]=0x08; out[13]=0x00;
  // IP header
  let p = 14;
  out[p]=0x45; out[p+1]=0; out.writeUInt16BE(0, p+2); // total len filled later
  out[p+8]=64; out[p+9]=17;                          // TTL, UDP
  ip2buf(GW).copy(out, p+12); ip2buf("255.255.255.255").copy(out, p+16);
  // UDP header
  let u = 34; out.writeUInt16BE(67, u); out.writeUInt16BE(68, u+2);
  // BOOTP/DHCP
  let d = 42;
  out[d]=2; out[d+1]=1; out[d+2]=6; out[d+3]=0;
  xid.copy(out, d+4);
  yourIp.copy(out, d+16);                            // yiaddr
  ip2buf(GW).copy(out, d+20);                        // siaddr
  chaddr.copy(out, d+28);
  out.writeUInt32BE(0x63825363, d+236);             // magic cookie
  let o = d + 240;
  const put = (code, ...bytes) => { out[o++]=code; out[o++]=bytes.length; for (const b of bytes) out[o++]=b; };
  put(53, replyType);
  put(54, ...ip2buf(GW));                            // server id
  put(51, (LEASE>>24)&255,(LEASE>>16)&255,(LEASE>>8)&255,LEASE&255); // lease
  put(1,  ...ip2buf(NETMASK));                       // subnet mask
  put(3,  ...ip2buf(GW));                            // router
  put(6,  ...ip2buf(DNS));                           // DNS
  out[o++]=255;                                      // end
  // lengths
  const udpLen = (o) - u; out.writeUInt16BE(udpLen, u+4);
  const ipTotal = (o) - 14; out.writeUInt16BE(ipTotal, 16);
  out.writeUInt16BE(checksum(out, 14, 34), 24);     // IP checksum
  safeSend(ws, out.slice(0, o));
  console.log(`DHCP ${msgType===1?"OFFER":"ACK"} -> ${cli.ip} (${mac2str(chaddr)})`);
  return true;
}

// ---- ARP responder (for the gateway) ----
function handleARP(ws, frame) {
  const ethType = (frame[12] << 8) | frame[13];
  if (ethType !== 0x0806) return false;
  const op = (frame[20] << 8) | frame[21];
  if (op !== 1) return false;                       // not request
  const targetIp = buf2ip(frame, 38);
  if (targetIp !== GW) return false;                // only answer for gateway
  const senderMac = frame.slice(22, 28);
  const senderIp = frame.slice(28, 32);
  const out = Buffer.alloc(42);
  senderMac.copy(out, 0); gwMac.copy(out, 6); out[12]=0x08; out[13]=0x06;
  out.writeUInt16BE(1, 14); out.writeUInt16BE(0x0800, 16); out[18]=6; out[19]=4;
  out.writeUInt16BE(2, 20);                          // reply
  gwMac.copy(out, 22); ip2buf(GW).copy(out, 28);
  senderMac.copy(out, 32); senderIp.copy(out, 38);
  safeSend(ws, out);
  return true;
}

// ---- main switch logic ----
const server = http.createServer(
  { maxHeaderSize: 16 * 1024, requestTimeout: 5000, headersTimeout: 5000 },
  (_, res) => { res.writeHead(200); res.end("BrowserOS relay\n"); }
);
const wss = new WebSocketServer({
  server,
  maxPayload: MAX_FRAME,
  perMessageDeflate: false,
  verifyClient(info, done) {
    if (wss.clients.size >= MAX_CLIENTS) {
      done(false, 503, "Relay full");
      return;
    }
    if (!trustedOrigin(info.req)) {
      done(false, 403, "Forbidden");
      return;
    }
    done(true);
  },
});

wss.on("connection", (ws) => {
  clients.set(ws, { mac: null, ip: null });
  console.log(`+ VM connected (${clients.size} total)`);

  ws.on("message", (data) => {
    const frame = Buffer.isBuffer(data) ? data : Buffer.from(data);
    if (frame.length < 14 || frame.length > MAX_FRAME) {
      ws.close(1009, "invalid frame size");
      return;
    }
    const src = mac2str(frame, 6);
    macTable.set(src, ws);
    const cli = clients.get(ws); if (cli) cli.mac = src;

    // Intercept services first.
    if (handleARP(ws, frame)) return;
    if (handleDHCP(ws, frame)) return;

    // Switch: forward to the destination port, or broadcast.
    const dst = mac2str(frame, 0);
    const bcast = dst === "ff:ff:ff:ff:ff:ff" || (frame[0] & 1);
    if (bcast) {
      for (const peer of wss.clients) if (peer !== ws) safeSend(peer, frame);
    } else {
      const target = macTable.get(dst);
      if (target) safeSend(target, frame);
      else for (const peer of wss.clients) if (peer !== ws) safeSend(peer, frame);
    }
    // NOTE: internet NAT (frames addressed to the gateway MAC bound for outside)
    // is NOT implemented here — see README. Inter-VM + DHCP fully work.
  });

  ws.on("close", () => {
    const cli = clients.get(ws);
    if (cli && cli.mac) macTable.delete(cli.mac);
    clients.delete(ws);
    console.log(`- VM disconnected (${clients.size} total)`);
  });
});

server.listen(PORT, HOST, () => {
  console.log(`BrowserOS virtual-switch relay on ws://${HOST}:${PORT}`);
  console.log(`LAN access: ${ALLOW_LAN} (set BROWSEROS_ALLOW_LAN=1 to opt in)`);
  if (ALLOWED_ORIGINS.size) console.log(`Extra allowed origins: ${[...ALLOWED_ORIGINS].join(", ")}`);
  console.log(`Subnet ${NET}.0/24  gateway ${GW}  (VMs get ${NET}.2, .3, ...)`);
  console.log(`In BrowserOS: Relay URL = ${ALLOW_LAN ? "ws://<this-host-private-ip>" : "ws://127.0.0.1"}:${PORT}`);
});

File: relay/relay.go
// BrowserOS Virtual Switch Relay (Go, single binary)
// ===========================================================================
// Same as relay.js: a WebSocket virtual L2 switch + DHCP that gives each
// browser-tab VM a UNIQUE IP and forwards Ethernet frames between them so they
// can ping/SSH each other.
//
// Build:  go build -o browseros-relay relay.go
// Run:    ./browseros-relay           (ws://127.0.0.1:9000)
// LAN:    BROWSEROS_ALLOW_LAN=1 ./browseros-relay
// In BrowserOS: Network -> Relay URL = ws://<host-ip>:9000
//
// Uses only the stdlib + a tiny inline WebSocket server (RFC6455, binary
// frames) so there are NO external dependencies.
package main

import (
	"bufio"
	"crypto/sha1"
	"encoding/base64"
	"encoding/binary"
	"fmt"
	"net"
	"net/url"
	"os"
	"strings"
	"sync"
	"time"
)

const (
	subnet        = "10.5.0"
	gateway       = subnet + ".1"
	maxFrameSize  = 65536
	maxHeaderSize = 16 << 10
	maxClients    = 64
)

var gwMAC = []byte{0x02, 0x00, 0x00, 0x00, 0x00, 0x01}

type client struct {
	conn net.Conn
	mac  string
	ip   string
	mu   sync.Mutex
}

var (
	mu       sync.Mutex
	clients  = map[*client]bool{}
	macTable = map[string]*client{}
	nextHost = 2
)

func allocIP() string {
	ip := fmt.Sprintf("%s.%d", subnet, nextHost)
	if nextHost >= 250 {
		nextHost = 2
	} else {
		nextHost++
	}
	return ip
}

func ip2b(s string) []byte {
	var b [4]byte
	parts := strings.Split(s, ".")
	for i := 0; i < 4 && i < len(parts); i++ {
		var v int
		fmt.Sscanf(parts[i], "%d", &v)
		b[i] = byte(v)
	}
	return b[:]
}
func macStr(b []byte) string {
	return fmt.Sprintf("%02x:%02x:%02x:%02x:%02x:%02x", b[0], b[1], b[2], b[3], b[4], b[5])
}
func ipChecksum(buf []byte) uint16 {
	var sum uint32
	for i := 0; i+1 < len(buf); i += 2 {
		sum += uint32(buf[i])<<8 | uint32(buf[i+1])
	}
	for sum>>16 != 0 {
		sum = (sum & 0xffff) + (sum >> 16)
	}
	return ^uint16(sum)
}

// ---- minimal WebSocket (server side, binary frames) ----
func wsAccept(key string) string {
	h := sha1.New()
	h.Write([]byte(key + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"))
	return base64.StdEncoding.EncodeToString(h.Sum(nil))
}
func wsWrite(c net.Conn, payload []byte) error {
	var hdr []byte
	n := len(payload)
	hdr = append(hdr, 0x82) // FIN + binary
	switch {
	case n < 126:
		hdr = append(hdr, byte(n))
	case n < 65536:
		hdr = append(hdr, 126, byte(n>>8), byte(n))
	default:
		hdr = append(hdr, 127, 0, 0, 0, 0, byte(n>>24), byte(n>>16), byte(n>>8), byte(n))
	}
	_, err := c.Write(append(hdr, payload...))
	return err
}
func wsRead(r *bufio.Reader) ([]byte, error) {
	b0, err := r.ReadByte()
	if err != nil {
		return nil, err
	}
	opcode := b0 & 0x0f
	if b0&0x80 == 0 {
		return nil, fmt.Errorf("fragmented websocket frames are unsupported")
	}
	b1, err := r.ReadByte()
	if err != nil {
		return nil, err
	}
	masked := b1&0x80 != 0
	if !masked {
		return nil, fmt.Errorf("client websocket frame is not masked")
	}
	ln := int(b1 & 0x7f)
	if ln == 126 {
		var e [2]byte
		if _, err = readFull(r, e[:]); err != nil {
			return nil, err
		}
		ln = int(binary.BigEndian.Uint16(e[:]))
	} else if ln == 127 {
		var e [8]byte
		if _, err = readFull(r, e[:]); err != nil {
			return nil, err
		}
		if e[0]|e[1]|e[2]|e[3] != 0 {
			return nil, fmt.Errorf("websocket frame too large")
		}
		ln = int(binary.BigEndian.Uint64(e[:]))
	}
	if ln > maxFrameSize {
		return nil, fmt.Errorf("websocket frame exceeds %d bytes", maxFrameSize)
	}
	var mask [4]byte
	if _, err = readFull(r, mask[:]); err != nil {
		return nil, err
	}
	data := make([]byte, ln)
	if _, err = readFull(r, data); err != nil {
		return nil, err
	}
	for i := range data {
		data[i] ^= mask[i%4]
	}
	switch opcode {
	case 0x2:
		return data, nil
	case 0x8:
		return nil, fmt.Errorf("close")
	default:
		return nil, fmt.Errorf("unsupported websocket opcode %d", opcode)
	}
}
func readFull(r *bufio.Reader, p []byte) (int, error) {
	got := 0
	for got < len(p) {
		n, err := r.Read(p[got:])
		got += n
		if err != nil {
			return got, err
		}
	}
	return got, nil
}

// ---- DHCP + ARP + switch ----
func (cl *client) send(b []byte) {
	cl.mu.Lock()
	defer cl.mu.Unlock()
	_ = cl.conn.SetWriteDeadline(time.Now().Add(2 * time.Second))
	err := wsWrite(cl.conn, b)
	_ = cl.conn.SetWriteDeadline(time.Time{})
	if err != nil {
		_ = cl.conn.Close()
	}
}

func handleDHCP(cl *client, f []byte) bool {
	if len(f) < 42 || f[12] != 0x08 || f[13] != 0x00 || f[23] != 17 {
		return false
	}
	ihl := int(f[14]&0x0f) * 4
	udp := 14 + ihl
	if udp+8 > len(f) {
		return false
	}
	dstPort := int(f[udp+2])<<8 | int(f[udp+3])
	if dstPort != 67 {
		return false
	}
	dhcp := udp + 8
	if dhcp+240 > len(f) {
		return false
	}
	xid := f[dhcp+4 : dhcp+8]
	chaddr := f[dhcp+28 : dhcp+34]
	msgType := byte(0)
	for o := dhcp + 240; o < len(f) && f[o] != 255; {
		if f[o] == 0 {
			o++
			continue
		}
		if o+1 >= len(f) {
			return false
		}
		code, l := f[o], int(f[o+1])
		if o+2+l > len(f) {
			return false
		}
		if code == 53 && l >= 1 {
			msgType = f[o+2]
		}
		o += 2 + l
	}
	if cl.ip == "" {
		mu.Lock()
		cl.ip = allocIP()
		mu.Unlock()
	}
	reply := byte(2)
	if msgType != 1 {
		reply = 5
	}
	out := make([]byte, 14+20+8+300)
	copy(out[0:], chaddr)
	copy(out[6:], gwMAC)
	out[12], out[13] = 0x08, 0x00
	p := 14
	out[p] = 0x45
	out[p+8] = 64
	out[p+9] = 17
	copy(out[p+12:], ip2b(gateway))
	copy(out[p+16:], []byte{255, 255, 255, 255})
	u := 34
	binary.BigEndian.PutUint16(out[u:], 67)
	binary.BigEndian.PutUint16(out[u+2:], 68)
	d := 42
	out[d], out[d+1], out[d+2] = 2, 1, 6
	copy(out[d+4:], xid)
	copy(out[d+16:], ip2b(cl.ip))
	copy(out[d+20:], ip2b(gateway))
	copy(out[d+28:], chaddr)
	binary.BigEndian.PutUint32(out[d+236:], 0x63825363)
	o := d + 240
	put := func(code byte, bytes ...byte) {
		out[o] = code
		out[o+1] = byte(len(bytes))
		o += 2
		copy(out[o:], bytes)
		o += len(bytes)
	}
	put(53, reply)
	put(54, ip2b(gateway)...)
	put(51, 0, 1, 0x51, 0x80) // ~86400s lease
	put(1, ip2b("255.255.255.0")...)
	put(3, ip2b(gateway)...)
	put(6, ip2b(gateway)...)
	out[o] = 255
	o++
	binary.BigEndian.PutUint16(out[u+4:], uint16(o-u))
	binary.BigEndian.PutUint16(out[16:], uint16(o-14))
	binary.BigEndian.PutUint16(out[24:], ipChecksum(out[14:34]))
	cl.send(out[:o])
	fmt.Printf("DHCP -> %s (%s)\n", cl.ip, macStr(chaddr))
	return true
}

func handleARP(cl *client, f []byte) bool {
	if len(f) < 42 || f[12] != 0x08 || f[13] != 0x06 {
		return false
	}
	if int(f[20])<<8|int(f[21]) != 1 {
		return false
	}
	target := fmt.Sprintf("%d.%d.%d.%d", f[38], f[39], f[40], f[41])
	if target != gateway {
		return false
	}
	out := make([]byte, 42)
	copy(out[0:], f[22:28])
	copy(out[6:], gwMAC)
	out[12], out[13] = 0x08, 0x06
	binary.BigEndian.PutUint16(out[14:], 1)
	binary.BigEndian.PutUint16(out[16:], 0x0800)
	out[18], out[19] = 6, 4
	binary.BigEndian.PutUint16(out[20:], 2)
	copy(out[22:], gwMAC)
	copy(out[28:], ip2b(gateway))
	copy(out[32:], f[22:28])
	copy(out[38:], f[28:32])
	cl.send(out)
	return true
}

func hostName(hostport string) string {
	if h, _, err := net.SplitHostPort(hostport); err == nil {
		return strings.ToLower(strings.Trim(h, "[]"))
	}
	return strings.ToLower(strings.Trim(hostport, "[]"))
}

func explicitlyAllowedOrigin(origin string) bool {
	for _, allowed := range strings.Split(os.Getenv("BROWSEROS_ALLOWED_ORIGINS"), ",") {
		if strings.TrimSpace(allowed) == origin {
			return true
		}
	}
	return false
}

func trustedOrigin(origin, requestHost string, allowLAN bool) bool {
	u, err := url.Parse(origin)
	if err != nil || (u.Scheme != "http" && u.Scheme != "https") || u.Host == "" {
		return false
	}
	if explicitlyAllowedOrigin(origin) {
		return true
	}
	if hostName(u.Host) != hostName(requestHost) {
		return false
	}
	host := strings.Trim(u.Hostname(), "[]")
	if host == "localhost" {
		return true
	}
	ip := net.ParseIP(host)
	return ip != nil && (ip.IsLoopback() || (allowLAN && (ip.IsPrivate() || ip.IsLinkLocalUnicast())))
}

func readHandshake(r *bufio.Reader, allowLAN bool) (string, error) {
	var key, host, origin, upgrade, connection, version string
	total := 0
	requestOK := false
	for {
		line, err := r.ReadString('\n')
		if err != nil {
			return "", err
		}
		total += len(line)
		if total > maxHeaderSize {
			return "", fmt.Errorf("websocket headers too large")
		}
		line = strings.TrimRight(line, "\r\n")
		if line == "" {
			break
		}
		if total == len(line)+2 || total == len(line)+1 {
			requestOK = strings.HasPrefix(line, "GET ") && strings.HasSuffix(line, " HTTP/1.1")
		}
		name, value, ok := strings.Cut(line, ":")
		if !ok {
			continue // request line
		}
		switch strings.ToLower(strings.TrimSpace(name)) {
		case "host":
			host = strings.TrimSpace(value)
		case "origin":
			origin = strings.TrimSpace(value)
		case "upgrade":
			upgrade = strings.TrimSpace(value)
		case "connection":
			connection = strings.TrimSpace(value)
		case "sec-websocket-key":
			key = strings.TrimSpace(value)
		case "sec-websocket-version":
			version = strings.TrimSpace(value)
		}
	}
	if !requestOK || key == "" || host == "" || origin == "" || version != "13" ||
		!strings.EqualFold(upgrade, "websocket") ||
		!strings.Contains(strings.ToLower(connection), "upgrade") ||
		!trustedOrigin(origin, host, allowLAN) {
		return "", fmt.Errorf("invalid or untrusted websocket upgrade")
	}
	return key, nil
}

func handleConn(conn net.Conn, allowLAN bool) {
	defer conn.Close()
	_ = conn.SetReadDeadline(time.Now().Add(5 * time.Second))
	r := bufio.NewReader(conn)
	key, err := readHandshake(r, allowLAN)
	if err != nil {
		conn.Write([]byte("HTTP/1.1 403 Forbidden\r\nConnection: close\r\n\r\n"))
		return
	}
	mu.Lock()
	full := len(clients) >= maxClients
	mu.Unlock()
	if full {
		conn.Write([]byte("HTTP/1.1 503 Service Unavailable\r\nConnection: close\r\n\r\n"))
		return
	}
	_ = conn.SetReadDeadline(time.Time{})
	resp := "HTTP/1.1 101 Switching Protocols\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Accept: " + wsAccept(key) + "\r\n\r\n"
	conn.Write([]byte(resp))

	cl := &client{conn: conn}
	mu.Lock()
	clients[cl] = true
	fmt.Printf("+ VM connected (%d total)\n", len(clients))
	mu.Unlock()
	defer func() {
		mu.Lock()
		if cl.mac != "" {
			delete(macTable, cl.mac)
		}
		delete(clients, cl)
		fmt.Printf("- VM disconnected (%d total)\n", len(clients))
		mu.Unlock()
	}()

	for {
		f, err := wsRead(r)
		if err != nil {
			return
		}
		if len(f) < 14 {
			continue
		}
		src := macStr(f[6:12])
		mu.Lock()
		macTable[src] = cl
		cl.mac = src
		mu.Unlock()

		if handleARP(cl, f) {
			continue
		}
		if handleDHCP(cl, f) {
			continue
		}

		dst := macStr(f[0:6])
		bcast := dst == "ff:ff:ff:ff:ff:ff" || f[0]&1 == 1
		mu.Lock()
		if !bcast {
			if t, ok := macTable[dst]; ok && t != cl {
				t.send(f)
				mu.Unlock()
				continue
			}
		}
		for peer := range clients {
			if peer != cl {
				peer.send(f)
			}
		}
		mu.Unlock()
	}
}

func main() {
	port := os.Getenv("PORT")
	if port == "" {
		port = "9000"
	}
	allowLAN := envBool("BROWSEROS_ALLOW_LAN")
	host := "127.0.0.1"
	if allowLAN {
		host = "0.0.0.0"
	}
	ln, err := net.Listen("tcp", net.JoinHostPort(host, port))
	if err != nil {
		fmt.Println("listen error:", err)
		os.Exit(1)
	}
	fmt.Printf("BrowserOS virtual-switch relay (Go) on ws://%s:%s\n", host, port)
	fmt.Printf("LAN access: %v (set BROWSEROS_ALLOW_LAN=1 to opt in)\n", allowLAN)
	fmt.Printf("Subnet %s.0/24  gateway %s  (VMs get %s.2, .3, ...)\n", subnet, gateway, subnet)
	if allowLAN {
		fmt.Printf("In BrowserOS: Relay URL = ws://<this-host-private-ip>:%s\n", port)
	} else {
		fmt.Printf("In BrowserOS: Relay URL = ws://127.0.0.1:%s\n", port)
	}
	for {
		conn, err := ln.Accept()
		if err != nil {
			continue
		}
		go handleConn(conn, allowLAN)
	}
}

func envBool(name string) bool {
	switch strings.ToLower(strings.TrimSpace(os.Getenv(name))) {
	case "1", "true", "yes", "on":
		return true
	default:
		return false
	}
}
